### PR TITLE
remote: honor --jobs for status collection and other fixes

### PR DESCRIPTION
**Pull request by Suor**

### CODE REVIEW

**Comment by efiop** [comment]
@Suor Just a heads up: tests failed.

**Comment by efiop** [comment]
@Suor ?

**Comment by Suor** [comment]
@efiop some ridiculous things a happening there. BTW, why we have this:
https://github.com/iterative/dvc/blob/432373fb42c7b2430cc2fc20b70d3dee67a4aaa7/dvc/repo/push.py#L7 
Looks like a bug.

**Comment by efiop** [comment]
@Suor Nice catch!

**Comment by Suor** [comment]
Looks like this fails for same reason @darabi had #2280.

**Comment by Suor** [comment]
@efiop This should fix a @darabi bug. Also, pretty commits here)

File: dvc/remote/ssh/connection.py [hunk 1]
```diff
@@ -307,18 +290,14 @@ def cp(self, src, dest):
         self.makedirs(posixpath.dirname(dest))
         self.execute("cp {} {}".format(src, dest))
 
-    @contextmanager
     def open_max_sftp_channels(self):
-        try:
-            channels = []
+        # If there are more than 1 it means we've already opened max amount
```
Comment (efiop) [inline dvc/remote/ssh/connection.py]: Could you provide a better explanation? Are we only able to open 1 sftp connection over 1 ssh connection? Why did this work before adding connection pool?

Comment (Suor) [reply dvc/remote/ssh/connection.py]: This the only place we are opening more that 1 sftp connection, so if we have more than 1 then we've already been here and opened max connections.

Before connection pool we were dropping `SSHConnection` after use and open a new one, so no "open max sftp - close all sftp - (reuse ssh conn) - try open sftp - Administratively prohobited" situation.

Comment (efiop) [reply dvc/remote/ssh/connection.py]: So why do we need _sftp_channels? Why not just `self.sftp`? If we are opening only 1 sftp channel per 1 ssh connection.

File: dvc/remote/ssh/connection.py [hunk 2]
```diff
@@ -181,52 +171,45 @@ def _remove_dir(self, path):
...
             self._remove_dir(path)
         else:
             self._remove_file(path)
 
     def download(self, src, dest, no_progress_bar=False, progress_title=None):
-        self._sftp_connect()
-
         if no_progress_bar:
-            self._sftp.get(src, dest)
+            self.sftp.get(src, dest)
```
Comment (efiop) [inline dvc/remote/ssh/connection.py]: why are we renaming all of these? is `sftp` used anywhere outside?

File: dvc/remote/ssh/connection.py [hunk 3]
```diff
@@ -61,26 +60,22 @@ def __init__(self, host, *args, **kwargs):
...
+        if not self._sftp_channels:
+            self._sftp_channels = [self._ssh.open_sftp()]
+        return self._sftp_channels[0]
 
     def close(self):
-        if self._sftp:
-            self._sftp.close()
-            self._sftp_alive = False
-
+        for sftp in self._sftp_channels:
```
Comment (efiop) [inline dvc/remote/ssh/connection.py]: If we now have only 1 connection, do we need this for loop at all?

File: dvc/remote/ssh/connection.py (hunk 1 above)
Comment (Suor) [reply dvc/remote/ssh/connection.py]: We are not opening 1 sftp per 1 connection. We are opening as max as we can, but only when we need them. Then we simply not close them and next time we need max channels they are already there.

Comment (efiop) [reply dvc/remote/ssh/connection.py]: Aaah, right. Got it.

**Comment by darabi** [comment]
I can confirm that when I install the current master the sftp connection exhaustion problems go away and I can `dvc push` and `dvc pull` via sftp.

Thanks

**Pull request merged**